### PR TITLE
Added .editorconfig files for Roslyn project

### PR DESCRIPTION
**Pull request by kuhlenh**
This is adding/updating editorconfig files in the repo to follow our code style guides when working inside the Roslyn project. 

There is a root editorconfig file that affects all Roslyn projects and the other editorconfig files tweak the rules for the IDE team folders specifically (only changes `var` preferences).

Tagging @dotnet/roslyn-ide and @MattGertz for approval.

### CODE REVIEW

File: src/EditorFeatures/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# CSharp IDE Team code style settings
```
Comment (jasonmalinowski) [inline src/EditorFeatures/.editorconfig]: This reads is the "C# IDE Team" is a thing.

File: src/EditorFeatures/.editorconfig [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+# Use var as much as possible
+[*.cs]
+csharp_style_var_for_locals = true:suggestion
```
Comment (jasonmalinowski) [inline src/EditorFeatures/.editorconfig]: Is this easier to define in the core one and then override in the compiler path?

File: .editorconfig [hunk 3]
```diff
@@ -24,3 +24,39 @@ indent_size = 2
...
+dotnet_style_qualification_for_event = false:suggestion
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
+dotnet_style_predefined_type_for_member_access = true:suggestion
+dotnet_style_object_initializer = true:suggestion
+dotnet_style_collection_initializer = true:suggestion
+dotnet_style_coalesce_expression = true:suggestion
+dotnet_style_null_propagation = true:suggestion
+
+# CSharp code style settings:
+# Only use var when it's obvious what the variable type is
```
Comment (jasonmalinowski) [inline .editorconfig]: Can we put the comments by the things actually doing that?

File: .editorconfig [hunk 4]
```diff
@@ -24,3 +24,39 @@ indent_size = 2
...
+# Prefer to generate C#6 and C#7 features but do not require
+[*.cs]
+csharp_style_var_for_locals = true:none
+csharp_style_var_for_built_in_types = true
```
Comment (tmat) [inline .editorconfig]: Not using keywords for built-in types such as string, int, etc?

File: .editorconfig (hunk 3 above)
Comment (kuhlenh) [inline .editorconfig]: yeah...do you want me to put a comment above each rule?

File: .editorconfig [hunk 5]
```diff
@@ -24,3 +24,39 @@ indent_size = 2
 # JSON files
 [*.json]
 indent_size = 2
+
+# Dotnet code style settings:
+# Avoid this. unless absolutely necessary
```
Comment (CyrusNajmabadi) [inline .editorconfig]: Avoid "this." unless necessary

File: .editorconfig (hunk 4 above)
Comment (CyrusNajmabadi) [inline .editorconfig]: true:none?

Comment (CyrusNajmabadi) [inline .editorconfig]: IDE uses 'var' whenever possible.

File: .editorconfig [hunk 6]
```diff
@@ -24,3 +24,39 @@ indent_size = 2
...
+
+# CSharp code style settings:
+# Only use var when it's obvious what the variable type is
+# Prefer to generate C#6 and C#7 features but do not require
+[*.cs]
+csharp_style_var_for_locals = true:none
+csharp_style_var_for_built_in_types = true
+csharp_style_var_when_type_is_apparent = true:none
+csharp_style_var_elsewhere = true:none
+csharp_style_expression_bodied_methods = false
```
Comment (CyrusNajmabadi) [inline .editorconfig]: false:none?

File: src/EditorFeatures/.editorconfig (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/.editorconfig]: Agreed.

**Comment by MattGertz** [comment]
(Kasey, could you tag me again once you have the 2 CRs -- thanks!)

File: .editorconfig (hunk 4 above)
Comment (kuhlenh) [inline .editorconfig]: @CyrusNajmabadi i was applying the "compiler rules" at the repo level and then have editorconfigs for the IDE projects overwrite these settings

File: .editorconfig (hunk 6 above)
Comment (kuhlenh) [inline .editorconfig]: yup thanks. I was playing around to see what @jmarolf implemented as defaults and apparently I forgot to change it back.

File: .editorconfig [hunk 7]
```diff
@@ -24,3 +24,39 @@ indent_size = 2
...
+csharp_style_expression_bodied_constructors = false:none
+csharp_style_expression_bodied_operators = false:none
+csharp_style_expression_bodied_properties = true:suggestion
```
Comment (kuhlenh) [inline .editorconfig]: (you recommended false:none last week...let's talk)

File: src/Features/.editorconfig [hunk 8]
```diff
@@ -0,0 +1,7 @@
+# CSharp IDE Team code style settings
```
Comment (CyrusNajmabadi) [inline src/Features/.editorconfig]: We don't need this file do we?

Comment (kuhlenh) [inline src/Features/.editorconfig]: oh shoot. I missed that one. Thanks!

File: .editorconfig [hunk 9]
```diff
@@ -24,3 +24,39 @@ indent_size = 2
 # JSON files
 [*.json]
 indent_size = 2
+
+# Dotnet code style settings:
+[*.{cs, vb}]
+# Avoid "this." unless absolutely necessary
```
Comment (CyrusNajmabadi) [inline .editorconfig]: avoid "this." when not necessary

**Comment by CyrusNajmabadi** [comment]
👍

File: .editorconfig (hunk 9 above)
Comment (CyrusNajmabadi) [inline .editorconfig]: avoid "this." or "me." when not necessary

**Comment by CyrusNajmabadi** [comment]
👍

**Comment by kuhlenh** [comment]
@MattGertz this has been reviewed by 2 people.

**Comment by MattGertz** [comment]
Approved for RC3 pending testing.

File: .editorconfig [hunk 10]
```diff
@@ -24,3 +24,45 @@ indent_size = 2
...
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
+dotnet_style_predefined_type_for_member_access = true:suggestion
+
+# Suggest more modern language features when available
+dotnet_style_object_initializer = true:suggestion
+dotnet_style_collection_initializer = true:suggestion
+dotnet_style_coalesce_expression = true:suggestion
+dotnet_style_null_propagation = true:suggestion
+
+# CSharp code style settings:
```
Comment (jasonmalinowski) [inline .editorconfig]: Seems a bit redundant: that's what the next line means. :-)

File: .editorconfig (hunk 9 above)
Comment (kuhlenh) [inline .editorconfig]: yes just realized i need to add the "Me." part. @Pilchie do I just add this and push again?

File: .editorconfig [hunk 11]
```diff
@@ -24,3 +24,45 @@ indent_size = 2
...
+
+# Dotnet code style settings:
+[*.{cs, vb}]
+# Avoid "this." unless absolutely necessary
+dotnet_style_qualification_for_field = false:suggestion
+dotnet_style_qualification_for_property = false:suggestion
+dotnet_style_qualification_for_method = false:suggestion
+dotnet_style_qualification_for_event = false:suggestion
+
+# Use language keywords instead of BCL types for type references
```
Comment (Pilchie) [inline .editorconfig]: I would say `framework type names` instead of `BCL types`.

File: src/Compilers/.editorconfig [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# Project-specific settings:
+# CSharp code style settings:
+[*.cs]
+csharp_style_var_for_built_in_types = false:none
+csharp_style_var_when_type_is_apparent = false:none
```
Comment (Pilchie) [inline src/Compilers/.editorconfig]: This should be `true:none`.

File: .editorconfig [hunk 13]
```diff
@@ -24,3 +24,45 @@ indent_size = 2
...
+dotnet_style_predefined_type_for_locals_parameters_members = true:suggestion
+dotnet_style_predefined_type_for_member_access = true:suggestion
+
+# Suggest more modern language features when available
+dotnet_style_object_initializer = true:suggestion
+dotnet_style_collection_initializer = true:suggestion
+dotnet_style_coalesce_expression = true:suggestion
+dotnet_style_null_propagation = true:suggestion
```
Comment (Pilchie) [inline .editorconfig]: Pre-emptively add `dotnet_style_explicit_tuple_names` from https://github.com/dotnet/roslyn/pull/15655.

**Comment by citizenmatt** [comment]
Is there any documentation for these settings available?

**Comment by jmarolf** [comment]
@citizenmatt working on writing the docs now.  currently the PR https://github.com/dotnet/roslyn/pull/15029 at least lists all the options but I don't have anything good to point to you at yet.

**Comment by citizenmatt** [comment]
Very cool, thanks! Are there any other options (planned or implemented) other than what's listed in #15029?

**Comment by jmarolf** [comment]
@citizenmatt There is also https://github.com/dotnet/roslyn/pull/15020 which will allow you to configure the IDE settings (this also needs some docs!).  Also working on merging in this change https://github.com/dotnet/roslyn/pull/15065 so you can configure naming styles in editorconfig.

**Comment by citizenmatt** [comment]
Brilliant. Thanks!

**Comment by jmarolf** [comment]
@citizenmatt got some documentation on these options now https://github.com/Microsoft/visualstudio-docs/pull/18.  Still more to do

**Pull request merged**